app/src/androidTest/java/org/mozilla/tv/firefox/FirefoxTestApplication.kt

```diff
@@ -0,0 +1,36 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox
+
+import android.arch.lifecycle.LiveData
+import android.arch.lifecycle.MutableLiveData
+import org.mozilla.tv.firefox.pocket.PocketEndpoint
+import org.mozilla.tv.firefox.pocket.PocketFeedStateMachine
+import org.mozilla.tv.firefox.pocket.PocketVideoRepo
+import org.mozilla.tv.firefox.pocket.PocketViewModel
+import org.mozilla.tv.firefox.utils.BuildConfigDerivables
+import org.mozilla.tv.firefox.utils.ServiceLocator
+
+class FirefoxTestApplication : FirefoxApplication() {
+
+    private val pocketEndpoint = object : PocketEndpoint() {
+        override suspend fun getRecommendedVideos(): List<PocketViewModel.FeedItem.Video>? {
+            return PocketViewModel.noKeyPlaceholders
+        }
+    }
+
+    private val pocketVideoRepoState = MutableLiveData<PocketVideoRepo.FeedState>()
+
+    private val pocketVideoRepo = object : PocketVideoRepo(pocketEndpoint, PocketFeedStateMachine(), BuildConfigDerivables()) {
+        override val feedState: LiveData<FeedState>
+            get() = pocketVideoRepoState
+    }
+
+    override fun createServiceLocator() = object : ServiceLocator(this) {
+        override val pocketRepo = pocketVideoRepo
+    }
+
+    fun pushPocketRepoState(state: PocketVideoRepo.FeedState) = pocketVideoRepoState.postValue(state)
```

Comment:
Our before block is called after the repo has already been injected into the VM, so updating the service locator at that point won't produce any changes.  We could do this kind of setup in a test's init block.  We aren't currently putting more than one UI test in each class, so they're essentially the same, but it is worrying to have that kind of land mine ("don't do setup in the standard JUnit setup block").  That said, you're right that otherwise it would be a much cleaner solution.

```
I/System.out: SEVTEST: FirefoxTestApplication#init                          
I/System.out: SEVTEST: FirefoxTestApplication#OnCreate                      
I/System.out: SEVTEST: PocketBasicUserFlowTest#init                         
I/System.out: SEVTEST: PocketViewModel#init (PocketVideoRepo instance set   
I/System.out: SEVTEST: PocketBasicUserFlowTest#@Before                      
I/System.out: SEVTEST: PocketBasicUserFlowTest#@Test                        
```

Reply:
I'm really surprised PocketBasicUserFlowTest runs before FirefoxTestApplication.onCreate! Unfortunately, I think that means that means there is no way for the test class to modify the application before init _except_ to specify a custom test runner. That'd look like:
```kotlin
// In same file; I think I'd prefer if these were inner classes to namespace but I don't know if it'd compile
class PocketBasicUserFlowTestApplication : FirefoxApplication() {
    override fun createServiceLocator() = object : ServiceLocator(this) {
        // override your stuff
    }
}

@RunWith(PocketBasicUserFlowTestRunner::class.java)
class PocketBasicUserFlowTest {
    // ...
}

class PocketBasicUserFlowTestRunner : TestApplicationRunner {
    override val testApplication = PocketBasicUserFlowTestApplication::class
}
```

I feel like I'd prefer that despite the boilerplate but I don't know how it looks in practice. I prefer it because I feel like the current TestApplication format seems like it's going to sprawl to handle the mutation of every possible state on the already initialized ServiceLocator: it'd be simple to customize the ServiceLocator for the current test only and do so without mutation.

Reply:
Turns out this runner isn't the same as the JUnit runners so this doesn't work.

We couldn't come up with something simple and it's probably not worth blocking right now but hopefully we can do something better later.

Reply:
We were unable to make this work.  Hopefully we can work a better solution out in the future.

From Slack:

> the @RunWith param is a class extending `Runner`, which doesn't have any fields related to `Application`
> we can provide an `Application` if we extend `AndroidJUnitRunner`, but (counterintuitively) that doesn't extend from `Runner`.  The only way I've found to set that class is through Gradle